### PR TITLE
[CARBONDATA-2752][CARBONSTORE] Carbon provide Zeppelin support

### PR DESCRIPTION
**Pull request by ajithme**
Apache Zeppelin is a popular open web-based notebook that enables interactive data analytics. This is one of the favored solutions for providing UI frontend as it can support solutions like Spark already. Carbon can leverage this to provide a UI for its operations. After CARBONDATA-2688 which provides a carbon REST server, we can add a UI support from zeppelin to provide a complete solution.

 - [ ] Document update required? YES, need to update usage guide for Zeppelin integration

 - [x] Testing done
1. Added UT
2. Done Testing manually by integration with Zeppelin
image

### CODE REVIEW

**Comment by ravipesala** [comment]
Can one of the admins verify this patch?

**Comment by CarbonDataQA** [comment]
Can one of the admins verify this patch?

File: store/sql/src/main/java/org/apache/carbondata/horizon/rest/controller/SqlHorizonController.java [hunk 1]
```diff
@@ -42,20 +43,24 @@
   public ResponseEntity<SqlResponse> sql(@RequestBody SqlRequest request) throws StoreException {
     RequestValidator.validateSql(request);
     List<Row> rows;
+    Dataset<Row> sqlDataFrame = null;
     try {
-      rows = SparkSqlWrapper.sql(SqlHorizon.getSession(), request.getSqlStatement())
+      sqlDataFrame = SparkSqlWrapper.sql(SqlHorizon.getSession(),
+              request.getSqlStatement());
+      rows = sqlDataFrame
           .collectAsList();
```
Comment (jackylk) [inline store/sql/src/main/java/org/apache/carbondata/horizon/rest/controller/SqlHorizonController.java]: move it to previous line

File: store/sql/src/main/java/org/apache/carbondata/horizon/rest/controller/SqlHorizonController.java [hunk 2]
```diff
@@ -42,20 +43,24 @@
...
     } catch (AnalysisException e) {
       throw new StoreException(e.getSimpleMessage());
     } catch (Exception e) {
       throw new StoreException(e.getMessage());
     }
-    Object[][] result = new Object[rows.size()][];
+    Object[][] result = new Object[rows.size()+1][];
```
Comment (jackylk) [inline store/sql/src/main/java/org/apache/carbondata/horizon/rest/controller/SqlHorizonController.java]: add space before and after `+`

File: store/sql/src/main/java/org/apache/carbondata/horizon/rest/controller/SqlHorizonController.java [hunk 3]
```diff
@@ -42,20 +43,24 @@
...
+    result[0] = sqlDataFrame.schema().fieldNames();
     for (int i = 0; i < rows.size(); i++) {
       Row row = rows.get(i);
-      result[i] = new Object[row.size()];
+      result[i+1] = new Object[row.size()];
       for (int j = 0; j < row.size(); j++) {
-        result[i][j] = row.get(j);
+        result[i+1][j] = row.get(j);
```
Comment (jackylk) [inline store/sql/src/main/java/org/apache/carbondata/horizon/rest/controller/SqlHorizonController.java]: Can't use System.arraycopy?

File: store/sql/src/main/java/org/apache/carbondata/horizon/rest/controller/SqlHorizonController.java [hunk 4]
```diff
@@ -42,20 +43,24 @@
...
           .collectAsList();
     } catch (AnalysisException e) {
       throw new StoreException(e.getSimpleMessage());
     } catch (Exception e) {
       throw new StoreException(e.getMessage());
     }
-    Object[][] result = new Object[rows.size()][];
+    Object[][] result = new Object[rows.size()+1][];
+    result[0] = sqlDataFrame.schema().fieldNames();
     for (int i = 0; i < rows.size(); i++) {
```
Comment (jackylk) [inline store/sql/src/main/java/org/apache/carbondata/horizon/rest/controller/SqlHorizonController.java]: `I ` can start from 1

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7295/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7297/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6064/

File: zeppelin/README.txt [hunk 5]
```diff
@@ -0,0 +1,18 @@
+Please follow below steps to integrate with zeppelin
```
[the remark this answers is not in the document]
Comment (ajithme) [reply zeppelin/README.txt]: Done:
https://github.com/ajithme/carbondata/tree/zeppelinsupport/integration/zeppelin

File: store/sql/src/main/java/org/apache/carbondata/horizon/rest/controller/SqlHorizonController.java (hunk 4 above)
Comment (ajithme) [reply store/sql/src/main/java/org/apache/carbondata/horizon/rest/controller/SqlHorizonController.java]: replaced with stream construct

File: store/sql/src/main/java/org/apache/carbondata/horizon/rest/controller/SqlHorizonController.java (hunk 3 above)
Comment (ajithme) [reply store/sql/src/main/java/org/apache/carbondata/horizon/rest/controller/SqlHorizonController.java]: Done

File: store/sql/src/main/java/org/apache/carbondata/horizon/rest/controller/SqlHorizonController.java (hunk 2 above)
Comment (ajithme) [reply store/sql/src/main/java/org/apache/carbondata/horizon/rest/controller/SqlHorizonController.java]: Done

File: store/sql/src/main/java/org/apache/carbondata/horizon/rest/controller/SqlHorizonController.java (hunk 1 above)
Comment (ajithme) [reply store/sql/src/main/java/org/apache/carbondata/horizon/rest/controller/SqlHorizonController.java]: Done

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7300/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6089/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7323/

File: zeppelin/README.txt (hunk 5 above)
Comment (jackylk) [inline zeppelin/README.txt]: Is this README written for carbon specificly?

File: zeppelin/assembly/assembly.xml [hunk 6]
```diff
@@ -0,0 +1,37 @@
+<assembly>
```
Comment (jackylk) [inline zeppelin/assembly/assembly.xml]: please move zeppelin folder to integration folder

File: zeppelin/src/main/java/org/apache/carbondata/zeppelin/CarbonInterpreter.java [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Carbon based interpreter for zeppelin
+ */
+public class CarbonInterpreter extends Interpreter {
+
+  public static final Logger logger = LoggerFactory.getLogger(CarbonInterpreter.class);
```
Comment (jackylk) [inline zeppelin/src/main/java/org/apache/carbondata/zeppelin/CarbonInterpreter.java]: Please use Carbon's LogService

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7352/

**Comment by ajithme** [comment]
Merged

**Pull request closed without merging**